trace/trace.go

```diff
@@ -0,0 +1,348 @@
+	if s.parentID != "" {
+		s.ev.AddField("trace.parent_id", s.parentID)
+	}
+	s.ev.AddField("trace.span_id", s.spanID)
+	// add rollup fields to the event
+	for k, v := range s.rollupFields {
+		s.ev.AddField(k, v)
+	}
+	for _, child := range s.children {
+		if !child.AmAsync() {
+			if !child.amFinished {
+				child.AddField("meta.finished_by_parent", true)
+				child.Finish()
+			}
+		}
+	}
+	s.amFinished = true
+	// if we're closing the root span, send the whole trace
+	if s.amRoot {
+		s.trace.Send()
+	}
+}
+
+// AmAsync reveals whether the span is asynchronous (true) or synchronous (false).
+func (s *Span) AmAsync() bool {
+	return s.amAsync
+}
+
+// GetChildren returns a list of all child spans (both synchronous and
+// asynchronous).
+func (s *Span) GetChildren() []*Span {
+	return s.children
+}
+
+// Get Parent returns this span's parent.
+func (s *Span) GetParent() *Span {
+	return s.parent
+}
+
+// ChildAsyncSpan creates a child of the current span that is expected to
+// outlive the current span (and trace). Async spans must be manually sent using
+// the `Send()` method but are otherwise identical to normal spans.
+func (s *Span) ChildAsyncSpan(ctx context.Context) (context.Context, *Span) {
+	ctx, newSpan := s.ChildSpan(ctx)
+	newSpan.amAsync = true
+	return ctx, newSpan
+}
+
+// Span creates a synchronous child of the current span. Spans must finish
+// before their parents.
+func (s *Span) ChildSpan(ctx context.Context) (context.Context, *Span) {
+	newSpan := newSpan()
+	newSpan.parent = s
+	newSpan.parentID = s.spanID
+	newSpan.trace = s.trace
+	newSpan.ev = s.trace.builder.NewEvent()
+	s.children = append(s.children, newSpan)
+	ctx = PutSpanInContext(ctx, newSpan)
+	return ctx, newSpan
+}
+
+// SerializeHeaders returns the trace ID, current span ID as parent ID, and an
+// encoded form of all trace level fields. This serialized header is intended to
+// be put in an HTTP (or other protocol) header to transmit to downstream
+// services so they may start a new trace that will be connected to this trace.
+// The serialized form may be passed to NewTrace() in order to create a new
+// trace that will be connected to this trace.
+func (s *Span) SerializeHeaders() string {
+	var prop = &propagation.Propagation{
+		TraceID:      s.trace.traceID,
+		ParentID:     s.spanID,
+		TraceContext: s.trace.traceLevelFields,
+	}
+	// prop.Source = HeaderSourceBeeline
+
+	return propagation.MarshalTraceContext(prop)
+}
+
+// Send sends this span and any synchronous span children. Does not send any
+// async children. Primarily used on asynchronous spans - you must call `Send()`
+// on an asynchronous span in order for it to get sent to Honeycomb. Normally,
+// this function is not used for synchronous spans; they all get sent when the
+// root span gets `Finish()`ed. While you can call `Send` on a synchronous
+// (normal) span, doing so prevents the span from getting any trace level fields
+// added after it is sent.
+func (s *Span) Send() {
```

Comment:
I think I want to leave the behavior as is for the time being and have updated the function description to be a little more explicit. I think it will be worth revisiting if we decide to go down the delayed send route for full traces.